Documents aggregate (#90)
HorizenLabs/zkverify-docs: docs/overview/04-proof-aggregation/02-handle-valid-proof.md

---
title: Handle Valid Proof
---

> When a valid proof is routed to a domain via a `domainId` provided in [`submitProof`](../02-mainchain/05-mainchain_api.md#submitproof) the proof is just added to the next aggregation in the given domain and some money are put on hold in the submitter wallet till the aggregation is published. The submitter hold money will be used to pay the aggregation's publisher: see [Publish Attestation](./03-publish-attestations.md) to learn more about the aggregation costs. The proof insertion in the aggregation emits a [`NewProof{statement, domainId, aggregationId}`](../02-mainchain/05-mainchain_api.md#newproof) event and the proof submitter should listen the events to wait a `NewAggregationReceipt{domainId, aggregationId, receipt}` event is emitted to match domain and aggregation: this step is important because the user can compute the Merkle path of its proof in the attestation only by **providing the block where this event was emitted**: the path can be computed by the [`aggregate_statementPath`](../02-mainchain/05-mainchain_api.md#aggregate_statementpath) RPC call.

Reviewer: ```suggestion
When a valid proof is routed to a domain via a `domainId` provided in [`submitProof`](../02-mainchain/05-mainchain_api.md#submitproof) the proof is just added to the next aggregation in the given domain and some money are put on hold in the submitter's wallet till the aggregation is published. The submitter's held money will be used to pay the aggregation's publisher: see [Publish Attestation](./03-publish-attestations.md) to learn more about the aggregation costs. The proof insertion in the aggregation emits a [`NewProof{statement, domainId, aggregationId}`](../02-mainchain/05-mainchain_api.md#newproof) event. The proof submitter should listen to a `NewAggregationReceipt{domainId, aggregationId, receipt}` event containing his domain and aggregation: this step is important because the user can compute the Merkle path of its proof in the attestation only by **providing the block where this event was emitted**: the path can be computed by the [`aggregate_statementPath`](../02-mainchain/05-mainchain_api.md#aggregate_statementpath) RPC call.
```

> That's the happy path, let's now take a look on what happen in the other paths. If something go wrong the **`submitProof` extrisic doesn't fail**, but just emit a [`CannotAggregate{statement, cause}`](../02-mainchain/05-mainchain_api.md#cannotaggregate) **event** with one of the following cause:

Reviewer: ```suggestion
That's the happy path, let's now take a look to what happens in the other situtations. If something goes wrong, the **`submitProof` extrisic doesn't fail**, but just emits a [`CannotAggregate{statement, cause}`](../02-mainchain/05-mainchain_api.md#cannotaggregate) **event** with one of the following causes:
```

- `DomainNotRegistered{domainId}`: if the submitter has provided an invalid `domainId`.

> - [`InvalidDomainState{domainId, state}`](./04-domain-management.md#remove-a-domain): the domain is not ready to accept any other proofs.

Reviewer: ```suggestion
- [`InvalidDomainState{domainId, state}`](./04-domain-management.md#remove-a-domain): the domain is not ready to accept any other proof.
```

- `DomainStorageFull{domainId}`: if the domain queue is full, in this case no proof can be added to that domain till at least an aggregation is published.

> - `InsufficientFunds`: if the submitter account hasn’t enough funds to cover its contribution for the aggregation's publication.

Reviewer: ```suggestion
- `InsufficientFunds`: if the submitter account hasn’t enough funds to cover for its contribution to the aggregation's publication.
```

So, to summarize, the submitter should listen the events generated by `submitProof` call and catch both:

- `NewProof` that is emitted if everything is ok and contains the aggregation coordinates
- `CannotAggregate` that is emitted if something has gone wrong and contains the cause of the failure

## Proof Submission Costs

> Every proof submitter in a given domain $D$ put the follow money amount $A$ on hold:

Reviewer: ```suggestion
Every proof submitter in a given domain $D$ puts the following money amount $A$ on hold:
```

$$
A(a) = \frac{Estimate(a) + Tip(a)}{size(a)}
$$

Where

- $a$ is the aggregation where the proof is inserted

> - $Estimate(a)$ is a price estimation on generate the aggregation **at the time of the submission**

Reviewer: ```suggestion
- $Estimate(a)$ is a price estimation for generating the aggregation **at the time of the submission**
```
